contribs/src/main/java/com/netflix/conductor/contribs/listener/ArchivingWithTTLWorkflowStatusListener.java

```diff
@@ -0,0 +1,114 @@
+
+    @Inject
+    public ArchivingWithTTLWorkflowStatusListener(ExecutionDAOFacade executionDAOFacade, Configuration config) {
+        this.executionDAOFacade = executionDAOFacade;
+        this.archiveTTLSeconds = config.getWorkflowArchivalTTL();
+        this.delayArchiveSeconds = config.getWorkflowArchivalDelay();
+
+        this.scheduledThreadPoolExecutor = new ScheduledThreadPoolExecutor(config.getWorkflowArchivalDelayQueueWorkerThreadCount(),
+                (runnable, executor) -> {
+                    LOGGER.warn("Request {} to delay archiving index dropped in executor {}", runnable, executor);
+                    Monitors.recordDiscardedArchivalCount();
+                });
+        this.scheduledThreadPoolExecutor.setRemoveOnCancelPolicy(true);
+    }
+
+    @PreDestroy
+    public void shutdownExecutorService() {
+        try {
+            LOGGER.info("Gracefully shutdown executor service");
+            scheduledThreadPoolExecutor.shutdown();
+            if (scheduledThreadPoolExecutor.awaitTermination(delayArchiveSeconds, TimeUnit.SECONDS)) {
+                LOGGER.debug("tasks completed, shutting down");
+            } else {
+                LOGGER.warn("Forcing shutdown after waiting for {} seconds", delayArchiveSeconds);
+                scheduledThreadPoolExecutor.shutdownNow();
+            }
+        } catch (InterruptedException ie) {
+            LOGGER.warn("Shutdown interrupted, invoking shutdownNow on scheduledThreadPoolExecutor for delay queue");
+            scheduledThreadPoolExecutor.shutdownNow();
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    @Override
+    public void onWorkflowCompleted(Workflow workflow) {
+        LOGGER.info("Archiving workflow {} on completion ", workflow.getWorkflowId());
+        if (delayArchiveSeconds > 0) {
+            scheduledThreadPoolExecutor.schedule(new DelayArchiveWorkflow(workflow, executionDAOFacade), delayArchiveSeconds, TimeUnit.SECONDS);
+        } else {
+            this.executionDAOFacade.removeWorkflowWithExpiry(workflow.getWorkflowId(), true, archiveTTLSeconds);
+            Monitors.recordWorkflowArchived(workflow.getWorkflowName(), workflow.getStatus());
+        }
+    }
+
+    @Override
+    public void onWorkflowTerminated(Workflow workflow) {
+        LOGGER.info("Archiving workflow {} on termination", workflow.getWorkflowId());
+        if (delayArchiveSeconds > 0) {
+            scheduledThreadPoolExecutor.schedule(new DelayArchiveWorkflow(workflow, executionDAOFacade), delayArchiveSeconds, TimeUnit.SECONDS);
+        } else {
+            this.executionDAOFacade.removeWorkflowWithExpiry(workflow.getWorkflowId(), true, archiveTTLSeconds);
+            Monitors.recordWorkflowArchived(workflow.getWorkflowName(), workflow.getStatus());
+        }
+    }
+
+    private class DelayArchiveWorkflow implements Runnable {
+        private final Workflow workflow;
```

Comment:
Since the removal of the workflow only uses the workflowId in the `run` method below in line 105, the object size can be reduced tremendously by only saving the worklfowId as part of the instance.

Reply:
Thanks for the suggestion. This will significantly reduce the memory consumption of the archival delay queue.